### PR TITLE
Add coordination logic into MI

### PR DESCRIPTION
**Pull request by madushadhanushka**
## Purpose
Adding clustering configurations into the MI

### Sample datasource configuration in deployment.toml
```
[[datasource]]
id = "WSO2_COORDINATION_DB"
url= "jdbc:mysql://localhost:3306/regdb"
username="root"
password="root"
driver="com.mysql.jdbc.Driver"
```

Carbon Coordination flow diagram

### CODE REVIEW

**Comment by Nirothipan** [comment]
@madushadhanushka Shall we fix the Codacy issues https://app.codacy.com/gh/wso2/micro-integrator/pullRequest?prid=5058203 and make it pass

File: components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java [hunk 1]
```diff
@@ -88,6 +97,22 @@ protected void activate(ComponentContext ctx) {
             bundleContext.registerService(TaskService.class.getName(), getTaskService(), null);
             //            bundleContext.registerService(Axis2ConfigurationContextObserver.class.getName(), new TaskAxis2ConfigurationContextObserver(getTaskService()), null);
 
+            CarbonDataSource dataSource = dataSourceService.getDataSource("WSO2_COORDINATION_DB");
```
Comment (Nirothipan) [inline components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java]: Shall we add a sample datasource config here

File: components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/query/QueryManager.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        DELETE_COORDINATOR,
+        DELETE_NODE_HEARTBEAT,
+        DELETE_REMOVED_MEMBER_DETAIL_FOR_NODE,
+        CLEAR_NODE_HEARTBEATS,
+        CLEAR_COORDINATOR_HEARTBEAT,
+        CLEAR_ALL_MEMBERSHIP_EVENTS,
+        CLEAN_MEMBERSHIP_EVENTS_FOR_NODE,
+    }
+
+    public enum DBTypes {
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/query/QueryManager.java]: What are the DBs we checked this with ?

Comment (Nirothipan) [reply components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/query/QueryManager.java]: Could you commit the scripts to be executed against the DB

Comment (Nirothipan) [reply components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/query/QueryManager.java]: We need to have the Db scripts ?

File: components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java (hunk 1 above)
Comment (madushadhanushka) [reply components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java]: I have added sample TOML config in the PR description

File: components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/query/QueryManager.java (hunk 2 above)
Comment (madushadhanushka) [reply components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/query/QueryManager.java]: I have tested this with MySQL only.

File: components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java (hunk 1 above)
Comment (Nirothipan) [reply components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java]: Shall we add the default data sources to toml , else adding this only will remove the rest

File: components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/RDBMSCommunicationBusContextImpl.java [hunk 3]
```diff
@@ -0,0 +1,1002 @@
...
+            close(resultSet, task);
+            close(preparedStatement, task);
+            close(clearMembershipEvents, task);
+            close(connection, task);
+        }
+    }
+
+    @Override
+    public String getCoordinatorNodeId(String groupId) throws ClusterCoordinationException {
+        {
```
Comment (arunans23) [inline components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/RDBMSCommunicationBusContextImpl.java]: Let's remove the additional brackets

File: components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java [hunk 4]
```diff
@@ -88,6 +97,22 @@ protected void activate(ComponentContext ctx) {
...
+                throw new DataSourceException("DataSource is not an RDBMS data source.");
+            }
+            RDBMSCoordinationStrategy rdbmsCoordinationStrategy = new
+                    RDBMSCoordinationStrategy((DataSource) coordinationDatasource);
+            rdbmsCoordinationStrategy.joinGroup();
+            List<NodeDetail> nodeDetailList = rdbmsCoordinationStrategy.getAllNodeDetails();
+            for (NodeDetail nodeDetail : nodeDetailList) {
+                log.info("Node connected: " + nodeDetail.getNodeId());
+            }
+
```
Comment (sdkottegoda) [inline components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java]: We need to do all this prior to registering the TaskService

File: components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/RDBMSCommunicationBusContextImpl.java [hunk 5]
```diff
@@ -0,0 +1,1002 @@
...
+            throw new ClusterCoordinationException("Error in connecting to the datasource to create tables " +
+                                                   "for clustering", e);
+        } finally {
+            close(connection, "Getting carbon coordination database information");
+        }
+
+        createTableIfNotExist(Table.LEADER_STATUS_TABLE);
+        createTableIfNotExist(Table.CLUSTER_NODE_STATUS_TABLE);
+        createTableIfNotExist(Table.MEMBERSHIP_EVENT_TABLE);
+        createTableIfNotExist(Table.REMOVED_MEMBERS_TABLE);
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.coordination/src/main/java/org/wso2/micro/integrator/coordination/RDBMSCommunicationBusContextImpl.java]: We will have to create the tables before hand and no need to check.

File: pom.xml [hunk 6]
```diff
@@ -1160,6 +1160,11 @@
                 <artifactId>opensaml</artifactId>
                 <version>${opensaml.orbit.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.wso2.ei</groupId>
+                <artifactId>org.wso2.micro.integrator.coordination</artifactId>
+                <version>1.2.0-SNAPSHOT</version>
```
Comment (Nirothipan) [inline pom.xml]: Let's not hard code the version

File: components/org.wso2.micro.integrator.ntask.core/pom.xml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.ntask.core/pom.xml]: Shall we add licence text

File: components/org.wso2.micro.integrator.ntask.core/pom.xml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        <version>1.2.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.wso2.micro.integrator.ntask.core</groupId>
+    <artifactId>org.wso2.micro.integrator.ntask.core</artifactId>
+    <version>1.2.0-SNAPSHOT</version>
+
+
+</project>
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.ntask.core/pom.xml]: We need a line here

File: components/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java [hunk 9]
```diff
@@ -0,0 +1,4 @@
+package org.wso2.micro.integrator.ntask.core.internal;
+
+public class TasksDSComponent {
+}
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java]: Why do we need this

File: components/org.wso2.micro.integrator.coordination/pom.xml [hunk 10]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.coordination/pom.xml]: Shall we add licence text

File: components/org.wso2.micro.integrator.coordination/pom.xml [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                        </Import-Package>
+                        <Export-Package>
+                            org.wso2.micro.integrator.coordination.*; version="${project.version}",
+                        </Export-Package>
+                    </instructions>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.coordination/pom.xml]: Let's add a line here

File: components/org.wso2.micro.integrator.coordination/pom.xml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <DynamicImport-Package>*</DynamicImport-Package>
+                        <Import-Package>
+                            *;resolution:=optional
```
Comment (Nirothipan) [inline components/org.wso2.micro.integrator.coordination/pom.xml]: Shall we add needed imports here

File: components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java [hunk 13]
```diff
@@ -82,12 +86,17 @@ protected void activate(ComponentContext ctx) {
             if (getTaskService() == null) {
                 taskService = new TaskServiceImpl();
             }
+
+            ClusterCoordinator clusterCoordinator = new ClusterCoordinator(dataSourceService);
+            clusterCoordinator.startCoordinator();
```
Comment (Nirothipan) [inline components/mediation/tasks/org.wso2.micro.integrator.ntask.core/src/main/java/org/wso2/micro/integrator/ntask/core/internal/TasksDSComponent.java]: We have to retrieve  CarbonDataSource dataSource = dataSourceService.getDataSource(RDBMSConstantUtils.COORDINATION_DB_NAME); and do this is only if it is valid.

**Comment by Nirothipan** [comment]
Can you please commit the scripts for db

File: features/org.wso2.micro.integrator.ntask.core.feature/pom.xml [hunk 14]
```diff
@@ -82,6 +86,7 @@
                                 <bundleDef>
                                     org.wso2.ei:org.wso2.micro.integrator.ntask.core
                                 </bundleDef>
+                                <bundleDef>org.wso2.ei:org.wso2.micro.integrator.coordination</bundleDef>
```
Comment (Nirothipan) [inline features/org.wso2.micro.integrator.ntask.core.feature/pom.xml]: It should be better to pack this in server feature

**Comment by sdkottegoda** [comment]
Merging the PR due to the urgency of its functionality. https://github.com/wso2/micro-integrator/issues/1208 for fixing code quality issues.

**Pull request merged**